### PR TITLE
WIP Separate the gts_ingest global and variable checks

### PR DESCRIPTION
**Pull request by daltonkell**
WIP to address #757 and #759

### CODE REVIEW

**Comment by daltonkell** [comment]
Pinging @benjwadams on this review. I tried to reuse some existing machinery for checking standard names and units of variables, but extracting it out of the CF module is a lift I'm not ready to undertake just yet (way outside the scope of this pull request or the issues it addresses). I'd like to know your opinion on the effectiveness of my choices:

https://github.com/ioos/compliance-checker/blob/e699f62e0d94644881dffffacbaab22371e951e4/compliance_checker/ioos.py#L967-L985

**Comment by daltonkell** [comment]
@mwengren I started working on #759 as it was basically in conjunction with #757. This is not in the milestone officially, but since #757 is, it wouldn't have made sense to not work on it. 

With these changes, testing against https://pae-paha.pacioos.hawaii.edu/erddap/tabledap/WQB-04.html (flagged for ingest) yields the following results:

```
...

gts_ingest requirements
* The following variables qualifed for GTS Ingest:
The following variables did not qualify for GTS Ingest:
 - oxygen_saturation
 - oxygen
 - turbidity
 - temperature
 - chlorophyll
 - salinity

...
```

These variables fail because they do not have an ancillary variable representing the QARTOD aggregate flag. What do you think so far?

**Comment by daltonkell** [comment]
Whoops -- I can't demote this to a draft PR, but it's still a draft PR. __NOT READY FOR MERGING!__

**Comment by mwengren** [comment]
@daltonkell Great, glad you are working on #759!  That will be a huge addition.

I see an issue here.  Looking only at the `temperature` variable on [WQB-04](https://pae-paha.pacioos.hawaii.edu/erddap/tabledap/WQB-04.das), I think it meets the requirements.  Checker should find the `ancillary_variables=temperature_qc_agg` (alongside many others), and see that is has a valid `standard_name=aggreagate_quality_flag`.  

Also, how is it determined which variables end up in the list of GTS-qualified|not qualified variables?  This is something I hadn't thought about.  With that test dataset, obviously there are many variables that are QC-related and not meant for publication on the GTS (they may be meant to be harvested by NDBC in connection to a data variable they describe, but they don't themselves qualify for GTS publication because they shouldn't have ancillary variables, etc).  

Seems the checker should output a list all variables that don't meet GTS requirements, even if they're QC-related.  So this should be a long list with all the QC-related variables tagged for not meeting GTS requirements.  Hopefully that's what happens here, can't quite tell.

**Comment by daltonkell** [comment]
Relating to the issue you are seeing -- very strange. I downloaded both the `.nc` and `.ncCF` files and `ncdump` -ed them, finding no `standard_name` of `aggregate_quality_flag`, but there it is in the `DAS`.

Which format would you typically use for testing?

**Comment by daltonkell** [comment]
> Also, how is it determined which variables end up in the list of GTS-qualified|not qualified variables?

I can use "meets requirements" instead of "qualified" if that's clearer. All this is doing is showing which variables met the requirements or not; I thought "qualified" would be more concise, but if it's actually referring to another categorization process, I'll change it to avoid any confusion.

**Comment by mwengren** [comment]
> I can use "meets requirements" instead of "qualified" if that's clearer. All this is doing is showing which variables met the requirements or not; I thought "qualified" would be more concise, but if it's actually referring to another categorization process, I'll change it to avoid any confusion.

Either wording is fine I think.  My question was more about how many variables in this particular dataset end up in the not qualified|not meeting requirements list.  Do all the QC-related variables end up being listed there?

> Relating to the issue you are seeing -- very strange. I downloaded both the .nc and .ncCF files and ncdump -ed them, finding no standard_name of aggregate_quality_flag, but there it is in the DAS.

> Which format would you typically use for testing?

Here's the command I ran for my own test of that dataset:
```
compliance-checker -t ioos -f html https://pae-paha.pacioos.hawaii.edu/erddap/tabledap/WQB-04 > WQB-04-ioos.html
```
I think your question gets back to discussion on #751.  I'd say `.ncCF` or `.ncCFMA` would be the right ones to pull out of ERDDAP.  

For WQB-04, the [.ncCF headers response](https://pae-paha.pacioos.hawaii.edu/erddap/tabledap/WQB-04.ncCFHeader?&time%3E=2020-01-06T16:00:00Z) contain those attributes, so files should as well.  

I'm not sure what CC is actually requesting under the hood, however!

**Comment by daltonkell** [comment]
Ahh ok I think I found the problem -- I downloaded the `.ncCF` file earlier, but it didn't request all the data! I re-downloaded it and here are the new results:

```
Running Compliance Checker on the datasets from: ['WQB-04.ncCF']
ut_scale(): NULL factor argument

--------------------------------------------------------------------------------
                         IOOS Compliance Checker Report                         
                                    ioos:1.2                                    
      https://ioos.github.io/ioos-metadata/ioos-metadata-profile-v1-2.html      
--------------------------------------------------------------------------------
                               Corrective Actions                               
WQB-04.ncCF has 28 potential issues


                               Highly Recommended                               
--------------------------------------------------------------------------------
cf_role
* attribute cf_role in variable platform1 not present

creator_country
* creator_country not present

creator_sector
* creator_sector not present

gts_ingest requirements
* The following variables qualifed for GTS Ingest:
 - turbidity
 - temperature
 - salinity
 - chlorophyll
 - oxygen
 - oxygen_saturation
The following variables did not qualify for GTS Ingest:
 - salinity_qc_din
 - turbidity_qc_spk
 - chlorophyll_qc_syn
 - chlorophyll_qc_nbr
 - oxygen_qc_gap
 - temperature_qc_loc
 - chlorophyll_qc_clm
 - oxygen_saturation_qc_agg
 - oxygen_saturation_qc_syn
 - salinity_qc_mvr
 - salinity_qc_syn
 - chlorophyll_qc_mvr
 - chlorophyll_qc_rtc
 - turbidity_qc_clm
 - oxygen_qc_rtc
 - oxygen_saturation_qc_rng
 - temperature_qc_rtc
 - temperature_qc_syn
 - temperature_qc_nbr
 - salinity_qc_agg
 - temperature_qc_rng
 - temperature_qc_flt
 - temperature_qc_din
 - salinity_qc_nbr
 - chlorophyll_qc_agg
 - oxygen_qc_syn
 - turbidity_qc_atn
 - oxygen_saturation_qc_loc
 - oxygen_saturation_qc_rtc
 - temperature_qc_crv
 - oxygen_qc_loc
 - chlorophyll_qc_atn
 - temperature_qc_agg
 - salinity_qc_atn
 - oxygen_qc_clm
 - turbidity_qc_syn
 - turbidity_qc_nbr
 - salinity_qc_loc
 - salinity_qc_rng
 - turbidity_qc_gap
 - oxygen_qc_rng
 - oxygen_saturation_qc_clm
 - oxygen_saturation_qc_atn
 - turbidity_qc_agg
 - temperature_qc_spk
 - temperature_qc_gap
 - turbidity_qc_loc
 - turbidity_qc_rng
 - temperature_qc_clm
 - chlorophyll_qc_gap
 - oxygen_qc_spk
 - chlorophyll_qc_rng
 - salinity_qc_clm
 - oxygen_qc_flt
 - oxygen_saturation_qc_gap
 - salinity_qc_rtc
 - oxygen_qc_agg
 - temperature_qc_mvr
 - salinity_qc_gap
 - oxygen_qc_atn
 - oxygen_saturation_qc_spk
 - salinity_qc_spk
 - oxygen_qc_mvr
 - oxygen_saturation_qc_flt
 - chlorophyll_qc_spk
 - chlorophyll_qc_flt
 - chlorophyll_qc_loc
 - turbidity_qc_mvr
 - oxygen_qc_nbr
 - turbidity_qc_flt
 - oxygen_saturation_qc_nbr
 - turbidity_qc_rtc
 - temperature_qc_atn
 - salinity_qc_flt
 - oxygen_saturation_qc_mvr
 - salinity_qc_crv

platform
* The global attribute "platform" must be a single string containing no blank characters; it is In Situ Ocean-based Platforms > Buoys, In Situ Ocean-based Platforms > Moorings

...
```

Obviously that first part about a `NULL factor argument` will be investigated, but I think this answers your question "will all the QC variables be included?"

**Comment by mwengren** [comment]
Yes, looks much better!  This will be a big help to our data providers.

File: compliance_checker/ioos.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
                                        get_instrument_variables,
                                        get_coordinate_variables)
 from compliance_checker import base
+from compliance_checker.cf import util as cf_util # not to be confused with cfutil.py
```
Comment (benjwadams) [inline compliance_checker/ioos.py]: This is confusing, but doesn't really merit a comment here -- maybe put it in the module and/or create an issue.

File: compliance_checker/ioos.py [hunk 2]
```diff
@@ -888,89 +889,163 @@ def check_platform_vocabulary(self, ds):
...
+        val (?): value to check
+
+        Returns
+        -------
+        bool
+        """
+
+        r = True
+        if isinstance(val, str):
+            if not (val.lower()=="true" or val.lower()=="false"):
```
Comment (benjwadams) [inline compliance_checker/ioos.py]: You could also use `val.lower() in {"true", "false"}`

File: compliance_checker/ioos.py [hunk 3]
```diff
@@ -888,89 +889,163 @@ def check_platform_vocabulary(self, ds):
...
+                r = False
+
+        elif (not isinstance(val, str)):
```
Comment (benjwadams) [inline compliance_checker/ioos.py]: This `isinstance` check is redundant given the preceding on in the `if` clause.

You could probably shorten this whole method by doing:

`return isinstance(val, str) and val.lower() in {"true", "false"}`

File: compliance_checker/ioos.py [hunk 4]
```diff
@@ -888,89 +889,163 @@ def check_platform_vocabulary(self, ds):
...
+        if do_gts: # execute if dataset flagged for ingest
+
+            var_passed_ingest_reqs = set()
+            var_passed_ingest_msg  = "The following variables qualifed for GTS Ingest:"
+            var_failed_ingest_msg  = "The following variables did not qualify for GTS Ingest:"
+
+            for v in ds.get_variables_by_attributes(gts_ingest=lambda x: x=="true"):
+                var_passed_ingest_reqs.add((v.name, self._var_qualifies_for_gts_ingest(ds, v)))
+
+            for vtup in var_passed_ingest_reqs:
```
Comment (benjwadams) [inline compliance_checker/ioos.py]: Use tuple unpacking for the arguments here with descriptive variable names.  I'm not sure what the elements of `vtup` are supposed to represent.

File: compliance_checker/ioos.py (hunk 1 above)
Comment (daltonkell) [reply compliance_checker/ioos.py]: I can create an issue, but the functionality is needed from the `cf/util.py` module, so at least this is unambiguously referring to that one and not `cfutil.py`. This PR isn't concerned with updating the CF modules.

File: compliance_checker/ioos.py (hunk 4 above)
Comment (daltonkell) [reply compliance_checker/ioos.py]: Good call here, descriptive variable names are always useful.

**Comment by daltonkell** [comment]
@benjwadams Feedback implemented

**Pull request merged**